### PR TITLE
ISPN-12531 Set default FD_SOCK ports and apply offset to them

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12531

### CODE REVIEW

**Comment by pruivo** [comment]
any reason to split the FD_SOCK port for UDP & TCP? you can use the same start port for both and it will pick the next available port. AFAICT, FD_SOCK broadcasts the port used (https://github.com/belaban/JGroups/blob/4.x/src/org/jgroups/protocols/FD_SOCK.java#L741)

**Comment by tristantarrant** [comment]
I used the same values that the old server and WildFly use. No other reason.

File: documentation/src/main/asciidoc/topics/ref_jgroups_preconfigured_tuning.adoc [hunk 1]
```diff
@@ -55,6 +55,12 @@ The following system properties apply to all JGroups stacks, including
 | 2000
 | Optional
 
+| `jgroups.fd.bind.port`
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/ref_jgroups_preconfigured_tuning.adoc:58]:
```suggestion
| `jgroups.fd.port`
```

File: documentation/src/main/asciidoc/topics/ref_jgroups_preconfigured_tuning.adoc [hunk 2]
```diff
@@ -55,6 +55,12 @@ The following system properties apply to all JGroups stacks, including
...
+| Bind port for the FD (failure detection protocol) socket.
+| `57600`
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/ref_jgroups_preconfigured_tuning.adoc:60]: put `54200` too?

File: server/runtime/src/main/java/org/infinispan/server/Bootstrap.java [hunk 3]
```diff
@@ -107,6 +107,12 @@ protected void handleArgumentCommand(String command, String parameter, Iterator<
             if (!properties.containsKey(Server.JGROUPS_BIND_PORT)) {
                properties.setProperty(Server.JGROUPS_BIND_PORT, Integer.toString(Server.DEFAULT_JGROUPS_BIND_PORT + offset));
             }
+            if (!properties.containsKey(Server.JGROUPS_FD_TCP_BIND_PORT)) {
```
Comment (pruivo) [inline server/runtime/src/main/java/org/infinispan/server/Bootstrap.java:110]: (I need to test this)
if I understand correctly, if the user sets `-Djgroups.fd.port=xxxx`  & `--offset=xxxx`, the `jgroups.fd.port` is ignored. is that the intention?

**Review by pruivo — CHANGES_REQUESTED** [review]

**Comment by pruivo** [comment]
using a single version would be simpler... anyway, I've made some comments :)

**Comment by pruivo** [comment]
@tristantarrant needs rebase.

File: server/runtime/src/main/java/org/infinispan/server/Bootstrap.java (hunk 3 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/Bootstrap.java:111]: No, if the user sets `-Djgroups.fd.port=xxxx` then it won't apply the offset

**Comment by tristantarrant** [comment]
> using a single port would be simpler... anyway, I've made some comments :)

I think @wfink 's opinion would be interesting to have here

**Comment by wfink** [comment]
yes, -o, -p and the -D*port* settings are conflicting and I would hope that no one would use that

But it would be good to have the same behavior for all the settings
server start parameter -p AND -Dinfinispan.bind.port should prefer "-p" but better show a WARN or reject
direct setting with server start parameter like -p or -D*port* should  be consistent with the -o offset
At the moment the combination
   bin/server.sh -o 100 -Djgroups.bind.port=8000 -p 22222
will use port 8000 (no offset) and 22322 (offset) - this is confusing if the jgroups ports use the settings differently

**Comment by tristantarrant** [comment]
@wfink the question was also related to why TCP and UDP use different ports. Maybe it makes sense to use just one

**Comment by wfink** [comment]
Also confusing is that all the JGroups ports use port-range>0 by default, so if the port is blocked it will simple increase the port# no matter whether offset or direct port is used, this can add confusion as well.
If we set the port-range=0 by default it would use the configured ports and it is ensured that the selected ports are used

File: server/runtime/src/main/java/org/infinispan/server/Bootstrap.java (hunk 3 above)
Comment (pruivo) [reply server/runtime/src/main/java/org/infinispan/server/Bootstrap.java:111]: no, the offset only checks `jgroups.fd.tcp.port` and `jgroups.fd.udp.port`, and It never checks `jgroups.fd.port`. And the configuration (`<FD_SOCK start_port="${jgroups.fd.tcp.port,jgroups.fd.port:57600}"/>`) first checks `jgroups.fd.tcp.port` and then `jgroups.fd.port`. One more reason I don't like to have multiple properties for the same xml attribute :(

I did the following test:
```
$ ./bin/server.sh -Djgroups.fd.port=9999
$ sudo ss -tlpn | grep pid=940614
LISTEN 0      4096            127.0.0.1:11222      0.0.0.0:*    users:(("java",pid=940614,fd=141))
LISTEN 0      50          192.168.2.100:7800       0.0.0.0:*    users:(("java",pid=940614,fd=106))
LISTEN 0      50          192.168.2.100:9999       0.0.0.0:*    users:(("java",pid=940614,fd=109))

$ ./bin/server.sh -Djgroups.fd.port=9999 -o 100
$ sudo ss -tlpn | grep pid=941806
LISTEN 0      4096            127.0.0.1:11322      0.0.0.0:*    users:(("java",pid=941806,fd=114))
LISTEN 0      50          192.168.2.100:7900       0.0.0.0:*    users:(("java",pid=941806,fd=106))
LISTEN 0      50          192.168.2.100:57700      0.0.0.0:*    users:(("java",pid=941806,fd=109))
```

File: documentation/src/main/asciidoc/topics/ref_jgroups_preconfigured_tuning.adoc (hunk 1 above)
Comment (pruivo) [reply documentation/src/main/asciidoc/topics/ref_jgroups_preconfigured_tuning.adoc]: the docs still show `jgroups.fd.bind.port`  which is incorrect.

File: server/runtime/src/main/java/org/infinispan/server/Bootstrap.java (hunk 3 above)
Comment (pruivo) [reply server/runtime/src/main/java/org/infinispan/server/Bootstrap.java:111]: and if you use `-Djgroups.fd.tcp.port=9999`, offset is ignored:

```
$ ./bin/server.sh -Djgroups.fd.tcp.port=9999 -o 100
$ sudo ss -tlpn | grep pid=942877
[sudo] password for pedro: 
LISTEN 0      4096            127.0.0.1:11322      0.0.0.0:*    users:(("java",pid=942877,fd=141))
LISTEN 0      50          192.168.2.100:7900       0.0.0.0:*    users:(("java",pid=942877,fd=106))
LISTEN 0      50          192.168.2.100:9999       0.0.0.0:*    users:(("java",pid=942877,fd=109))
```

File: documentation/src/main/asciidoc/topics/ref_jgroups_system_properties.adoc [hunk 4]
```diff
@@ -55,6 +55,11 @@ Set system properties that configure JGroups cluster transport stacks.
 | 10000
 | Optional
 
+| `jgroups.fd.bind.port`
```
Comment (wfink) [inline documentation/src/main/asciidoc/topics/ref_jgroups_system_properties.adoc:58]: If I look to the changes I see separate jgroups.fd.tcp.port and jgroups.fd.udp.port and jgroups.fd.port
Seems jgroups.fd.bind.port is wrong for documentation

**Review by wfink — CHANGES_REQUESTED** [review]

**Comment by wfink** [comment]
I also noticed that the FD_SOCK (tcp) port default is different in comment and real files

"TCP FD port is 57800" but the files use 57600

**Comment by tristantarrant** [comment]
I have rebased this with the following changes:
* use 57600 for FD, regardless of TCP/UDP stacks
* I use a single `jgroups.fd.port` property everywhere

**Review by pruivo — APPROVED** [review]

**Pull request merged**